### PR TITLE
Add Dockerfile-source to Knative Service Pipelines

### PR DESCRIPTION
**Pull request by navidshaikh**
# Changes
- Uses `buildah` task for Dockerfile-source at git to build and push
- Uses `kn` task to deploy as Knative Service
- Documents creating `docker-registry` type secrets, for push/pull
- Documents configuring `ServiceAccount` for deploying ksvc and linking the secrets to it
- Adds Pipelines for updating a ksvc as well as performing some traffic operations
- Links the example Pipelines from kn task README

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [ ] Yaml file complies with [yamllint](https://github.com/adrienverge/yamllint) rules.

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: kn/knative-dockerfile-deploy/README.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+```bash
+kubectl create namespace tkn-kn
+```
+
+2 - Create a `docker-registry` type secrets for pushing/pulling the built container images.
+```bash
+kubectl create secret docker-registry container-registry --docker-server=<DOCKER-REGISTRY> --docker-username=<USERNAME> --docker-password=<PASSWORD> --docker-email=<EMAIL>
+```
+#### Note:
+- If you are using `docker.io`: Push to (a new private repo at) `docker.io` via buildah and pull via `imagePullSecrets` does not work!
```
Comment (chmouel) [inline kn/knative-dockerfile-deploy/README.md]: I am not sue i understand this part : 

pulll via `imagePullSecrets`

maybe i misread it,

File: kn/knative-dockerfile-deploy/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  Please [create a new](https://hub.docker.com/repository/create) empty public repository for this tutorial and refer it in subsequent steps.
+- If you are using `quay.io`: Push to `quay.io` via buildah and pull via `imagePullSecrets` works well. [Get](https://quay.io/signin/) an account created at quay.
```
Comment (chmouel) [inline kn/knative-dockerfile-deploy/README.md]: If you wanted to tell the user to have a default `imagePullSecrets`maybe you want to link this https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/ in there?

File: kn/knative-dockerfile-deploy/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+- kind: ServiceAccount
+  name: kn-deployer-account
+  namespace: tkn-kn
+roleRef:
+  kind: ClusterRole
+  name: kn-deployer
+  apiGroup: rbac.authorization.k8s.io
+```
+
+  - If you've used the same names for namespace and secrets as mentioned above, you can configure the ServiceAccount with YAML file in this repo using:
```
Comment (chmouel) [inline kn/knative-dockerfile-deploy/README.md]: with *the* YAML from this repository

File: kn/knative-dockerfile-deploy/build_deploy/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
+# Build and Deploy Pipeline
+
+Let's create a container image from a git source having a Dockerfile and deploy it to a Knative Service.
+This Pipeline builds and push the git source using `buildah` and deploys the container as Knative Service using `kn`.
+
+## Pipline:
```
Comment (chmouel) [inline kn/knative-dockerfile-deploy/build_deploy/README.md]: Pipeline

File: kn/knative-dockerfile-deploy/README.md (hunk 1 above)
Comment (navidshaikh) [reply kn/knative-dockerfile-deploy/README.md]: If we push to a non-existing repo at docker.io via buildah task, it doesn't populate the repoName at docker hub, then, referencing the same repo from docker.io in kn task can't pull the image.
This works though if I pull via podman CLI locally. 
image

I had a draft issue against catalog buildah task for this, then I discarded this as this doesn't seem to be the case if pushed in same setup to quay.io. May be something on dockerhub side..

Comment (chmouel) [reply kn/knative-dockerfile-deploy/README.md]: I think it just need a bit of rephrasing,

**Comment by chmouel** [comment]
It would be nice to start to think abotu adding a test, I know it's not going to be trivial tho.....

File: kn/knative-dockerfile-deploy/README.md (hunk 1 above)
Comment (navidshaikh) [reply kn/knative-dockerfile-deploy/README.md]: Rephrased to

- For using `docker.io`: Please [create a new](https://hub.docker.com/repository/create) empty public repository and refer it in subsequent steps.
- For using `quay.io`: No need to create a repository beforehand.

File: kn/knative-dockerfile-deploy/README.md (hunk 2 above)
Comment (navidshaikh) [reply kn/knative-dockerfile-deploy/README.md]: Sure, added a line in above section where secrets are created as:

 More about secrets for [`interacting with a private registry`](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).

File: kn/knative-dockerfile-deploy/README.md (hunk 3 above)
Comment (navidshaikh) [reply kn/knative-dockerfile-deploy/README.md]: updated.

File: kn/knative-dockerfile-deploy/build_deploy/README.md (hunk 4 above)
Comment (navidshaikh) [reply kn/knative-dockerfile-deploy/build_deploy/README.md]: updated.

**Comment by navidshaikh** [comment]
sorry for following up late on this :-/

> It would be nice to start to think abotu adding a test, I know it's not going to be trivial tho.....

@chmouel : tekton e2e tests are added in kn's repo https://github.com/knative/client/blob/master/test/tekton-tests.sh by @mgencur. 
These tests run in a nightly job, which installs tektoncd/pipelines, and uses `buildah`, `kn` tasks
from catalog to build helloworld and deploy using kn's nightly build.
The catch here is though, these tekton e2e tests are ran after the kn's nightly build, so its non-blocking.

To add e2e here in catalog we'd need to configure serving, deps etc. WDYT ? I can look at adding e2e for kn task as well these pipelines (may be a separate PR?).

**Pull request merged**